include/tvm/te/operation.h

```diff
@@ -76,6 +78,11 @@ class TVM_DLL OperationNode : public Object {
    * \return type of i-th output.
    */
   virtual DataType output_dtype(size_t i) const = 0;
+  /**
+   * @brief Returns memory scope of operation
+   * TODO(amalyshe): add support for individual output's tensors, not the only one
+   */
+  String memory_scope() const { return memory_scope_; }
```

Comment:
Memory scope is a part of the TIR Buffer and can be communicated via the tensor to buffer `binds` map provided when lowering the TE tensors. See https://github.com/apache/tvm/blob/main/src/relay/backend/te_compiler.cc#L346. As you can see, this binds field is empty and is not being used by the TE Compiler today in main. Instead of annotating the placeholders with scope from the virtual device, we should use the virtual device to construct the TIR buffers with the appropriate scope and store them in the cached func to be used in the binds field when LowerSchedule is called by the te_compiler.

Reply:
Done